source/mir/random/tinflex/internal/types.d

```diff
@@ -0,0 +1,318 @@
+        static if (isFloatingPoint!S)
+        {
+            alias seq =  AliasSeq!(x, c, tx, t1x, t2x);
+            foreach (i, v; seq)
+            {
+                assert(!v.isNaN, "variable " ~ seq[i].stringof ~ " isn't allowed to be NaN");
+            }
+        }
+    }
+}
+
+/// ditto
+IntervalPoint!S intervalPoint(F0, F1, F2, S)
+                             (in F0 f0, in F1 f1, in F2 f2, in S x, in S c = 1.5)
+    if (is(ReturnType!F0 == S) && is(ReturnType!F1 == S) && is(ReturnType!F2 == S) &&
+        (isFloatingPoint!S))
+{
+    return intervalPoint(f0(x), f1(x), f2(x), x, c);
+}
+
+IntervalPoint!S intervalPoint(S)
+                             (in S f0, in S f1, in S f2, in S x, in S c = 1.5)
+    if (isFloatingPoint!S)
+{
+    return IntervalPoint!S(f0, f1, f2, x, c);
+}
+
+/**
+Notations of different function types according to the Tinflex paper.
+It is based on this naming scheme:
+
+- a: concAve
+- b: convex
+- Type 4 is the pure case without any inflection point
+*/
+enum FunType {T1a, T1b, T2a, T2b, T3a, T3b, T4a, T4b}
+
+/**
+Checks whether the function type is concave
+*/
+bool isConcave(in FunType t) pure @safe @nogc nothrow
+{
+    with(FunType)
+    return t == T1a || t == T2a || t == T3a || t == T4a;
+}
+
+/**
+Checks whether the function type is convex
+*/
+bool isConvex(FunType t) pure @safe @nogc nothrow
+{
+    return !isConcave(t);
+}
+
+/**
+Determine the function type of an interval.
+Based on Theorem 1 of the Tinflex paper.
+Params:
+    bl = left side of the interval
+    br = right side of the interval
+*/
+FunType determineType(F0, F1, F2, S)
+                     (in F0 f0, in F1 f1, in F2 f2, in S bl, in S br)
+    if (is(ReturnType!F0 == S) && is(ReturnType!F1 == S) && is(ReturnType!F2 == S) &&
+        (isFloatingPoint!S))
+{
+    return determineType(intervalPoint(f0, f1, f2, bl), intervalPoint(f0, f1, f2, br));
+}
+
+/// ditto
+FunType determineType(S)(in IntervalPoint!S l, in IntervalPoint!S r)
+    if (isFloatingPoint!S)
+{
+    import std.traits : isFloatingPoint;
+
+    static if (isFloatingPoint!S)
+    {
+        import std.math : isInfinity, isNaN;
+    }
+    else
+    {
+        alias isInfinity = (x) => false;
+        alias isNaN = (x) => false;
+    }
+
+
+    assert(l.x < r.x, "invalid interval");
+    assert(!isNaN(l.tx) && !isNaN(r.tx), "Invalid interval points");
+
+    // slope
+    auto R = (r.tx - l.tx) / (r.x- l.x);
+
+    with(FunType)
+    {
+        import std.stdio;
+        //TODO: This needs more tests
+        if (isInfinity(l.tx))
+        {
+            writeln("infA");
+            if (r.t2x < 0 && r.t1x > 0)
+                return T4a;
+            else
+                assert(0, "Invalid case");
+        }
+
+        if (isInfinity(r.tx))
+        {
+            writeln("infB");
+            if (l.t2x < 0 && l.tx > 0)
+                return T4a;
+            else
+                assert(0, "Invalid case");
+        }
+
+        if ((l.c > 0 && r.tx == 0.0) ||
```

Comment:
Do not need to handle +-infinity here. +- infinity should be replaced with +-S.max in the before iteration